core/src/main/java/com/linecorp/armeria/server/AbstractVirtualHostBuilder.java

```diff
@@ -283,75 +284,193 @@ B service(T serviceWithPathMappings, Function<T, R> decorator) {
      * Binds the specified annotated service object under the path prefix {@code "/"}.
      */
     public B annotatedService(Object service) {
-        return annotatedService("/", service);
+        return annotatedService("/", service,
+                                ImmutableList.of(), ImmutableList.of(), ImmutableList.of(),
+                                Function.identity());
     }
 
     /**
      * Binds the specified annotated service object under the path prefix {@code "/"}.
      */
-    public B annotatedService(
-            Object service,
-            Function<Service<HttpRequest, HttpResponse>,
-                     ? extends Service<HttpRequest, HttpResponse>> decorator) {
-        return annotatedService("/", service, decorator);
+    public B annotatedService(Object service,
+                              List<ExceptionHandlerFunction> exceptionHandlers) {
+        return annotatedService("/", service,
+                                exceptionHandlers, ImmutableList.of(), ImmutableList.of(),
+                                Function.identity());
     }
 
     /**
      * Binds the specified annotated service object under the path prefix {@code "/"}.
      */
-    public B annotatedService(Object service, Map<Class<?>, ResponseConverter> converters) {
-        return annotatedService("/", service, converters);
+    public B annotatedService(Object service,
+                              List<RequestConverterFunction> requestConverters,
+                              List<ResponseConverterFunction> responseConverters) {
+        return annotatedService("/", service,
+                                ImmutableList.of(), requestConverters, responseConverters,
+                                Function.identity());
     }
 
     /**
      * Binds the specified annotated service object under the path prefix {@code "/"}.
      */
-    public B annotatedService(
-            Object service, Map<Class<?>, ResponseConverter> converters,
-            Function<Service<HttpRequest, HttpResponse>,
-                     ? extends Service<HttpRequest, HttpResponse>> decorator) {
-        return annotatedService("/", service, converters, decorator);
+    public B annotatedService(Object service,
+                              List<ExceptionHandlerFunction> exceptionHandlers,
+                              List<RequestConverterFunction> requestConverters,
+                              List<ResponseConverterFunction> responseConverters) {
```

Comment:
I feel like the overloaded `annotatedService()` methods are now quite confusing because they look pretty close to each other. How about just accepting `Iterable<?>` and `Object...`, and then doing `instanceof` and downcasts? i.e.

```java
public B annotatedService(Object service, Object... exceptionHandlersAndConverters) {
    return annotatedService(service, ImmutableList.copyOf(requireNonNull(exceptionHandlersAndConverters)));
}

public B annotatedService(Object service, Iterable<?> exceptionHandlersAndConverters) {
    for (Object o : exceptionHandlersAndConverters) {
        boolean added = false;
        if (o instanceof ExceptionHandler) {
            exceptionHandlers.add((ExceptionHandlerFunction) o);
        }
        if (o instanceof RequestConverterFunction) {
            requestConverters.add((RequestConverterFunction) o);
        }
        if (o instanceof ResponseConverterFunction) {
            responseConverters.add((ResponseConverterFunction) o);
        }
        if (!added) {
            throw new IllegalArgumentException(...);
        }
    }
    ...
}
```

Please note that I did not use `else if` when checking the type of an object. This allows a user pass an object that implements more than one interface, e.g.

```java
public class MyAnnotatedServiceExtension implements ExceptionHandlerFunction, RequestConverterFunction, ResponseConverterFunction { ... }
```

However, as you know, the 3 interfaces all have the method `accept` which will be confusing when a class implements all of them. Therefore, we may need to rename `accept()` to `canHandle(Throwable)` and `canConvert(...)`. What do you think? Do you think it is desirable for a user to implement the 3 interfaces in a single class for his/her convenience?